src/cpu/aarch64/acl_post_ops.cpp

```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
+* Copyright 2022 Arm Ltd. and affiliates
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "cpu/aarch64/acl_gemm_convolution.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
+namespace aarch64 {
+
+status_t acl_post_ops_t::execute(const exec_ctx_t &ctx, void *src_orig) const {
+
+    int post_op_index = 0;
+    int sum_index = pd()->acp.sum_index;
+
+    // As these are post ops, this src will also be our dst. If we have a sum
+    // post op, the src/dst will start off in a temporary, then change to
+    // DNNL_ARG_DST after the sum.
+    void *src = src_orig;
```

Comment:
Hi @jondea, I'm glad it performs now at the level you expect.
We agreed inside the team that current solution is a bit more complex than it might be.
We ask you to follow the design approach implemented for reference cpu: [here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/primitive_attr_postops.hpp) and [here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/primitive_attr_postops.cpp).
It should be enough to fulfil the purpose and simple enough as it allows not to touch main library infrastructure.
Could you proceed with it, please? Thank you.

Reply:
Hi @dzarukin.

Unfortunately we can’t follow the structure of `ref_post_ops_t` completely.

One reason is that `ref_post_ops_t` applies each post op in sequence on each element as a scalar. This makes the use of SIMD difficult and introduces a lot of branching. To make use of SIMD and move the branching to a higher level we apply each post op completely on the whole tensor, fusing if we can.

`acl_post_ops_t` makes use of `acl_binary_t` and `acl_eltwise_fwd_t`, so its structure needs to reflect those primitives. Specifically, we need a `pd_t`-like struct with an `init` (which returns success/unimplemented), a constructor from that struct along with `create_resource` and `execute` methods.

But I think we can get closer. Does the following work? Change `acl_post_ops_t` so that it no longer inherits from `primitive_t`, and instead just implements the methods it needs. This would allow us to get rid of any changes to the main library infrastructure because we would no longer need a new primitive kind.

Reply:
Hi @jondea, thank you for the comment and sorry for a delayed answer.

It should be fine to have a class that uses methods as implementations and follows same steps of creation and initialization.
Since it has to have `init`, you might want to keep it as a part of `pd_t` on implementation side.
The way execute function will work for ACL is implementation defined. I bet you can implement it the way to utilize vectorization by applying full post-op first.
v3.0 will have some dramatical changes on the core architecture side, thus, you might not want to post-op class having dependencies on it, and it seems it's not needed.
Feel free to ask for help if needed. Thank you.

Reply:
No problem, thanks for the response.

Just to check, do you mean that the `acl_post_ops_t::pd_t` should still inherit from `primitive_desc_t` but `acl_post_ops_t` shouldn't inherit from `primitive_t`? Or have I misunderstood?

When you refer to changes in v3.0 which may affect this PR, do you mean #1391 and to a lesser extent #1260? Or are there any other relevant changes to be aware of?

Reply:
Hi @jondea 

> Just to check, do you mean that the acl_post_ops_t::pd_t should still inherit from primitive_desc_t but acl_post_ops_t shouldn't inherit from primitive_t? Or have I misunderstood?

I meant the following:
```
struct some_impl_t : public primitive_t {
    struct pd_t : public cpu_some_pd_t {
        ...
        status_t init(engine_t *engine) {
            ...
            CHECK(po = std::make_shared<acl_post_ops_t>(...));
            CHECK(po->init());
        }
        ...
        std::shared_ptr<acl_post_ops_t> po; // A place to store post_ops object.
    }
```

Create resource part accesses `po->create_resource(...)`, execute part accesses `po->execute(...)`.

> When you refer to changes in v3.0 which may affect this PR, do you mean https://github.com/oneapi-src/oneDNN/pull/1391 and to a lesser extent https://github.com/oneapi-src/oneDNN/pull/1260? Or are there any other relevant changes to be aware of?

Mostly first one. I don't know all the details about implementation but can imagine how pieces are moved and changed significantly. Thank you.

Reply:
Thanks for the clarification and code snippet, that helped me understand. I have made these changes, let me know if they look okay.